### ISSUE TITLE
Clang does not diagnose an explicit specialization or partial specialization of initializer_list

### ISSUE THREAD

**Issue by miyuki** [llvm/llvm-project#126270]
According to the C++ standard (C++23 [initializer.list.syn] paragraph 2):

> If an explicit specialization or partial specialization of `initializer_list` is declared, the program is ill-formed.

It does not say "no diagnostic is required"; therefore, I think Clang must produce an error, but it does not. Example code:

```
#include <initializer_list>

template<>
class std::initializer_list<int> {};

template<typename T>
class std::initializer_list<T*> {};
```

Clang (current trunk version) accepts the code with `-std=c++23 -pedantic-errors`.

**Comment by miyuki** [comment]
I noticed that Clang now has a `no_specializations` attribute, so it should probably be easier to implement this diagnostic in libc++ rather than in the frontend.

**Comment by shafik** [comment]
CC @ldionne @mordante this looks like a libc++ issue, correct?

**Comment by mordante** [comment]
Thanks for catching this!

This is [LWG2129: User specializations of std::initializer_list](https://wg21.link/LWG2129) which has been closed as completed by 3687d3c2e9e1d4dcfb76e7d7614413eede5af951. Both GCC and MSVC have implemented this issue in the compiler. Using the `no_specializations` attributes solves it in libc++. This does not fix the issue for clang using libstdc++.

Since libstdc++ is the default standard library on Linux I feel it should be fixed in the compiler.

@ldionne do we want to also fix this in libc++ using the `no_specializations` attribute? (If so I can create a patch.)
